
#Càrrega de les llibreries necessàries
# Llibreria bs4dash desconfigua el panell de l'esquerra
library(shiny)
library(shinydashboard)
library(DT)
library(shinycssloaders)
library(shinyjs)
library(dplyr)

ui <- dashboardPage(
  
  skin="purple",
  
  dashboardHeader(title = "GenoView"),
  
  #Sidebar content
  dashboardSidebar(
    sidebarMenu(
      id = "tabName",
      menuItem("Home", tabName = "inici", icon = icon("home")),
      menuItem("Study Searcher", tabName = "searcher", icon = icon("search")),
      menuItem("Quality Control of Raw Data", tabName = "quality_control", icon = icon("chart-line")),
      conditionalPanel(
        condition = "input.tabName === 'searcher'",
        textInput("studyID", "Study ID:"),
        textInput("keyword", "Keyword in Abstract:"),
        textInput("runAccession", "Run Accession:"),
        textInput("specie", "Specie:"),
        textInput("year", "Year:"),
      ),
      conditionalPanel(
        condition = "input.tabName === 'quality_control'",
        textInput("qcRunAccession", "Run Accession for QC:")
      )
    )
  ),
  
  #Body content
  dashboardBody(
    tabItems(
      tabItem(tabName = "inici",
              #Informació sobre la App i sobre el repositori SRA que es mostra a la pestanya inicial
              h1("GenoView"),
              p("This application aims to be a search engine for scientific studies addressed to researchers and the entire scientific community. The application is based on the SRA data repository, showing the studies contained in this database, allowing researchers to filter and select studies of their interest to subsequently perform a quality control of the raw data and, if they are interested, download the data in fastq format for subsequent omics data analysis studies."),
              h3("Information about Sequence Read Archive (SRA)"),
              p("SRA is a public repository maintained by the National Center for Biotechnology Information (NCBI) at the National Institutes of Health (NIH) in the United States. The SRA acts as a hub for the storage of high-throughput sequencing data generated by various sequencing technologies."),
              p("The SRA primarily stores short read sequences obtained from next-generation sequencing (NGS) experiments. NGS technologies allow researchers to sequence large amounts of DNA or RNA in parallel, generating massive data sets. The SRA serves as a crucial resource for researchers to access and download these sequencing reads for various subsequent analyses."),
              h3("Types of data analysis that SRA information supports"),
              p("SRA data can be used for a wide range of bioinformatics analyses, such as:"),
              p("- Genome assembly and annotation: Reconstructing the complete DNA sequence of an organism or identifying functional elements within a genome."),
              p("- Transcriptome analysis: Studying gene expression patterns by analyzing RNA sequencing data."),
              p("- Variant search: Identification of genetic variations such as single nucleotide polymorphisms (SNPs) or insertions/deletions (indels) within a population."),
              p("- Metagenomic and microbiome analysis: Characterization of the composition of microbial communities present in a sample."),
              h3("Advantages of using SRA"),
              p("- Centralized repository: Provides a single access point to a vast collection of sequencing data."),
              p("- Standardized formats: Data is stored in standardized formats, ensuring accessibility and interoperability."),
              p("- Free and open access: SRA data is available for free download and reuse by the scientific community."),
              h2("Using StudySeekerNGS"),
              p("The application has two main tabs: Study Searcher and Quality Control:"),
              p("The Study Searcher tab allows you to view experiments and studies from the SRA database. It provides search and filter options across various text fields, enabling you to select studies of interest. The studies selected in this tab will automatically appear in the Quality Control tab."),
              p("The Quality Control tab displays the selected studies along with a button. Upon clicking this button, the associated files of the experiment are downloaded, and a quality control check is performed. The result is an HTML report displaying the quality graphs of the file."),
              p("Additionally, this tab provides an option to manually enter the Run Accession of the experiment of interest and perform a quality analysis."),
              p("This way, StudySeeker offers a seamless workflow from study selection to quality control, making it a comprehensive tool for scientific study exploration and analysis.")
              
      ),
      
      #Pestanya 2: taula amb els estudis i cercador
      tabItem(tabName = "searcher",
              fluidRow(
                box(title = "Results", width = 12, 
                    shinycssloaders::withSpinner(DT::dataTableOutput("taula"))),
                ),
              tabsetPanel(
                tabPanel("Runs per year", plotOutput("run_histogram")),
                tabPanel("Studies per year", plotOutput("study_histogram"))
                )
              ),
      
      #Pestanya 3: control de qualitat
      tabItem(tabName = "quality_control",
              h2("Quality Control of Raw Data"),
              p("The quality control of raw data is generated with the fastqc command, resulting in an HTML file. To perform the quality control of the data from the selected study, enter the Run Accession in the search box of the left side panel."),
              h3("The following details the information obtained from quality controls with this tool:"),
              p("- Basic statistics: a table indicating the file type, sequencing method, total number of sequences, sequences marked as poor quality, sequence length, and GC percentage."),
              p("- Sequence quality: sequences located at the top of the graph (green color) are assumed to be of good quality, while sequences located in the middle and lower part (yellow and red colors, respectively) are considered of average or poor quality. The average quality is represented with a blue line."),
              p("- Distribution of sequence quality scores: represents the average quality score over the length of all reads (X-axis) and the total number of reads with this score (Y-axis)."),
              p("- Content of each base in the sequences: representation of the percentage of each of the four nucleotides (T, C, G, A) at each position in all reads of the input sequence file."),
              p("- GC content per sequence: shows the number of reads obtained vs. the percentage of G and C bases per read. It is compared with a theoretical distribution that assumes a uniform GC content for all reads, where the central peak corresponds to the global GC content of the underlying genome."),
              p("- Base N content: if a sequencer is not able to make a base call with enough confidence, it writes an “N” instead of a conventional base call."),
              p("- Sequence length distribution: distribution of fragment sizes in the analyzed file. If they are variable length fastq files, the relative quantities of each different sequence fragment size are shown."),
              p("- Sequence duplication levels: the total percentage of reads of a certain sequence in the file is shown in purple. A low level of duplication may indicate a very high level of coverage of the target sequence, while a high level of duplication is more likely to indicate some type of enrichment bias."),
              p("- Overrepresented sequences: finding that a single sequence is highly over-represented in the set means that it is highly biologically significant or that the library is contaminated or not as diverse as expected."),
              p("- Adapter content: shows the accumulated percentage of reads with the different adapter sequences at each position. Once an adapter sequence is seen in a read, it is counted as present in the final end, so the percentage increases as the read length does."),
              useShinyjs(),
    
              actionButton("runQC", "Run Quality Control"),
              #div(id = "loading", style = "display: none;", withSpinner(div())),  # Spinner to be shown during loading
              DT::dataTableOutput("selected_studies"),
              tabsetPanel(
                tabPanel("QC Results", withSpinner(htmlOutput("qc_report")))
                
              )
      )
    )
  )
)
